multilinear_extensions/src/virtual_polys.rs

```diff
@@ -184,155 +196,139 @@ impl<'a, E: ExtensionField> VirtualPolynomials<'a, E> {
     ///
     /// the per-thread instances are registered locally and stored in `thread_based_mles_storage`
     /// using the MLE’s raw pointer as the key to ensure uniqueness and reference consistency.
-    pub fn register_mles(&mut self, mles: Vec<&'a ArcMultilinearExtension<'a, E>>) {
+    pub fn register_mles(&mut self, mles: Vec<EitherRefMLE<'a, E>>) -> Vec<usize> {
         let log2_num_threads = log2_strict_usize(self.num_threads);
+        let mut indexes = vec![];
         for mle in mles {
-            let mle_ptr: usize = Arc::as_ptr(mle) as *const () as usize;
-            let mles = (0..self.num_threads)
-                .map(|thread_id| {
-                    let mle_thread_based = if mle.num_vars() > log2_num_threads {
-                        self.get_range_polys_by_thread_id(thread_id, vec![mle])
-                            .remove(0)
+            let poly_meta = if mle.num_vars() > log2_num_threads {
+                PolyMeta::Normal
+            } else {
+                PolyMeta::Phase2Only
+            };
+            let mles = match mle {
+                Either::Left(mle) => {
+                    (0..self.num_threads)
+                        .map(|thread_id| {
+                            let mle_thread_based = if mle.num_vars() > log2_num_threads {
+                                self.get_subslice_polys_by_thread_id(thread_id, vec![mle])
+                                    .remove(0)
+                            } else {
+                                // polynomial is too small
+                                Arc::new(mle.as_view_slice(1, 0))
+                            };
+                            mle_thread_based
+                        })
+                        .collect_vec()
+                }
+                Either::Right(mle) => {
+                    if mle.num_vars() > log2_num_threads {
+                        mle.as_view_chunks_mut(self.num_threads)
+                            .into_iter()
+                            .map(Arc::new)
+                            .collect_vec()
                     } else {
-                        // polynomial is too small
-                        Arc::new(mle.get_ranged_mle(1, 0))
-                    };
-                    self.polys[thread_id].register_mle(mle_thread_based.clone());
-                    mle_thread_based
-                })
-                .collect_vec();
-            self.thread_based_mles_storage.insert(mle_ptr, mles);
+                        vec![mle.as_owned(); self.num_threads]
+                            .into_iter()
+                            .map(Arc::new)
+                            .collect_vec()
+                    }
+                }
+            };
+            let index = self
+                .polys
+                .iter_mut()
+                .zip_eq(mles)
+                .map(|(poly, mle)| poly.register_mle(mle))
+                .collect_vec()
```

Comment:
how about we add more sanity check like assert the indices are equal?

Reply:
Since modified register_mles to private in https://github.com/scroll-tech/ceno/pull/940/commits/9e2c7feb6ed0f0ac7f188fdceb440a7fb468763b, I would prefer avoid massive sanity check on the function associate with it.